tests/integration/framework/process/sqlite/sqlite.go

```diff
@@ -0,0 +1,147 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sqlite
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"path/filepath"
+	"strconv"
+	"testing"
+
+	// Blank import for the sqlite driver
+	_ "modernc.org/sqlite"
+
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	commonapi "github.com/dapr/dapr/pkg/apis/common"
+	componentsv1alpha1 "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
+)
+
+// Option is a function that configures the process.
+type Option func(*options)
+
+// SQLite database that can be used in integration tests.
+type SQLite struct {
+	dbPath string
+	conn   *sql.DB
+	opts   options
+}
+
+func New(t *testing.T, fopts ...Option) *SQLite {
+	t.Helper()
+
+	s := &SQLite{
+		opts: options{
+			name:     "mystore",
+			metadata: map[string]string{},
+			connect:  true,
+		},
+	}
+
+	for _, fopt := range fopts {
+		fopt(&s.opts)
+	}
+
+	// Create a SQLite database in the test's temporary directory
+	tmpDir := t.TempDir()
+	s.dbPath = filepath.Join(tmpDir, "test-data.db")
+	// t.Logf("Storing SQLite database at %s", s.dbPath)
+
+	// Connect to the database
+	// We do this here because the test's init code may need to access the database before starting the test
+	if s.opts.connect {
+		var err error
+		s.conn, err = sql.Open("sqlite", "file://"+s.dbPath+"?_txlock=immediate&_pragma=busy_timeout(5000)&_pragma=journal_mode(WAL)")
+		require.NoError(t, err, "Failed to connect to SQLite database")
+	}
+
+	return s
+}
+
+func (s *SQLite) Run(t *testing.T, ctx context.Context) {
+	// Nothing here
+}
+
+func (s *SQLite) Cleanup(t *testing.T) {
+	if s.conn != nil {
+		err := s.conn.Close()
+		require.NoError(t, err)
+	}
+}
+
+// GetConnection returns the connection to the SQLite database.
+// This will be nil if the "connect" is false.
+func (s *SQLite) GetConnection() *sql.DB {
+	return s.conn
+}
+
+// CreateStateTables creates the state tables in a new SQLite DB.
+func (s *SQLite) CreateStateTables(t *testing.T) {
+	require.NotNil(t, s.conn, "Cannot perform if connection isn't active")
+	_, err := s.conn.Exec(`
+CREATE TABLE metadata (
+  key text NOT NULL PRIMARY KEY,
+  value text NOT NULL
+);
+INSERT INTO metadata VALUES('migrations','1');
+CREATE TABLE state (
+  key TEXT NOT NULL PRIMARY KEY,
+  value TEXT NOT NULL,
+  is_binary BOOLEAN NOT NULL,
+  etag TEXT NOT NULL,
+  expiration_time TIMESTAMP DEFAULT NULL,
+  update_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
+);`)
+	require.NoError(t, err)
```

Comment:
It should only be used in some edge cases, such as when the database needs to be pre-populated with data.

Reply:
No, we can't do that in `Run` because it's executed in parallel with daprd starting. This must be done before daprd starts up

Reply:
Sorry I misspoke. Now that I've more time I went back and looked again at what the issue was.

We must create the tables before calling `Run`, because we must write data in the database before calling `Run`. Otherwise, we'd have to populate the database in the **test's** `Run` method, but at that point daprd is already running.

Reply:
Daprd is not running at this point.

Reply:
When the test is Run-ing, daprd is not??

Reply:
I don't think there's anything I can do here as we need to run the queries before daprd starts.

Regardless of whether processes are started in sequence or parallel, the test's `Run` method is invoked after all processes have started, so we cannot execute queries to write data in that method.

Please confirm and I'll resolve the conversation.

Reply:
Sorry, I don't know what you mean and think there is some confusion on the intention of a `suite.Case`, the framework `Process` and the order things are executed.

The test `Case` is the main test definition, whose struct name and package tree derive the Go test name. Its `Setup` function is always run first, and is where test processes (test dependencies) are defined. Next, the framework will execute the `Run` function of the test case process dependencies in order. This is where the process executes the actual dependency- this could be a binary process, in memory server, writing to a database w.e.
The test `Case` Run is then executed where the test invocation and checking code lives. Finally the processes are then `Cleanup`ed in reverse order to the `Run` order- it is important that processes ensure all resources are released here.

Below is the patch file for these changes showing how the integration test process is designed to be structured.

[sqlite.patch.txt](https://github.com/dapr/dapr/files/13599591/sqlite.patch.txt)